### PR TITLE
add attemptNarrow to ApplicativeErrorOps

### PR DESCRIPTION
**Pull request by marcodippy**
Hello, I already copy-pasted this function in a few places in my codebases, so I thought it could be useful to have it in cats.

Example use case (assuming `F : ApplicativeError[F, Throwable]`):
```scala
def authenticate(r: Request[F]): F[Unit] = ???

val authentication: Kleisli[F, Request[F], Either[AuthenticationException, Unit]] =
  Kleisli { request => authenticate(request).attemptCase[AuthenticationException] }
```

### CODE REVIEW

File: core/src/main/scala/cats/syntax/applicativeError.scala [hunk 1]
```diff
@@ -83,6 +86,9 @@ final class ApplicativeErrorOps[F[_], E, A](private val fa: F[A]) extends AnyVal
   def attempt(implicit F: ApplicativeError[F, E]): F[Either[E, A]] =
     F.attempt(fa)
 
+  def attemptCase[EE](implicit F: ApplicativeError[F, E], tag: ClassTag[EE], ev: EE <:< E): F[Either[EE, A]] =
+    F.recover(F.map(fa)(_.asRight[EE])) { case e: EE => e.asLeft[A] }
```
Comment (diesalbla) [inline core/src/main/scala/cats/syntax/applicativeError.scala]: Given that extra inter-dependencies could lead to circular dependencies, perhaps we can by-pass the `either` syntax and use this not much longer alternative: 
  ```suggestion
      F.recover(F.map(fa)(x => Right[EE, A](x))) { case e: EE => Left[EE, A](e) }
```

**Comment by kailuowang** [comment]
I think this could be useful although I am not quite sure about the name. I am thinking `attemptNarrow` or `narrowAttempt`?

**Comment by marcodippy** [comment]
@kailuowang renamed to `attemptNarrow` 👍

File: core/src/main/scala/cats/syntax/applicativeError.scala [hunk 2]
```diff
@@ -83,6 +85,9 @@ final class ApplicativeErrorOps[F[_], E, A](private val fa: F[A]) extends AnyVal
   def attempt(implicit F: ApplicativeError[F, E]): F[Either[E, A]] =
     F.attempt(fa)
 
+  def attemptNarrow[EE](implicit F: ApplicativeError[F, E], tag: ClassTag[EE], ev: EE <:< E): F[Either[EE, A]] =
```
Comment (tkroman) [inline core/src/main/scala/cats/syntax/applicativeError.scala]: why do we need a classtag here?

Comment (marcodippy) [reply core/src/main/scala/cats/syntax/applicativeError.scala]: `EE` would be erased and we wouldn't be able to pattern match on it otherwise @tkroman

Comment (tpolecat) [reply core/src/main/scala/cats/syntax/applicativeError.scala]: It's unsound without `ev` due to erasure, but `ev` is unused in the body so we might want to add a comment to this effect.

**Comment by marcodippy** [comment]
@tpolecat I added the test, let me know what you think

**Comment by marcodippy** [comment]
hey @kailuowang, is there anything I can do to get this merged?

**Comment by kailuowang** [comment]
Hi @marcodippy sorry about the delay here, but can I ask one more thing? Now that we dropped 2.11 from master branch, we can add this method to the typeclass itself and have the syntax delegate to it. Would you  make that change?

**Comment by marcodippy** [comment]
no problem @kailuowang, I just pushed the change :)

**Pull request merged**